File: Haskell/ssort.hs
-- Selection sort.
--
-- Skupoy Sergey.  sergey.archlinux@gmail.com, 2015
--
split :: Ord a => [a] -> a -> [a] -> [a]
split [] m r     = m : (ssort r)
split (x:xs) m r = case x < m of
                       True  -> split xs x (m:r)
                       False -> split xs m (x:r)

ssort :: Ord a => [a] -> [a]
ssort []     = []
ssort (x:xs) = split xs x []

File: Clojure/ssort.clj
;; Selection sort
;;
;; Skupoy Sergey. sergey.archlinux@gmail.com, 2015.

(defn ssort [arr]
  (if (empty? arr)
    []
    (_split (rest arr) (first arr) [])))

(defn _split [arr key acc]
  (loop [a arr k key ac acc]
    (if (empty? a)
      (cons k (ssort ac))
      (if (< (first a) k)
        (recur (rest a) (first a) (cons k ac))
        (recur (rest a) k (cons (first a) ac))))))




   



